File: src/features/typing/data/go/median_of_two_sorted_arrays.go
package main

import "math"

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	m, n := len(nums1), len(nums2)
	if m > n {
		return findMedianSortedArrays(nums2, nums1)
	}
	total := m + n
	half := total / 2
	left, right := 0, m
	for left <= right {
		i := left + (right-left)/2
		j := half - i
		aLeft := minValue(nums1, i)
		aRight := maxValue(nums1, i, m)
		bLeft := minValue(nums2, j)
		bRight := maxValue(nums2, j, n)
		if aLeft <= bRight && bLeft <= aRight {
			if total%2 == 1 {
				return float64(min2(aRight, bRight))
			}
			return float64(max2(aLeft, bLeft)+min2(aRight, bRight)) / 2
		}
		if aLeft > bRight {
			right = i - 1
		} else {
			left = i + 1
		}
	}
	return 0
}

func minValue(nums []int, index int) int {
	if index == 0 {
		return math.MinInt
	}
	return nums[index-1]
}

func maxValue(nums []int, index int, size int) int {
	if index == size {
		return math.MaxInt
	}
	return nums[index]
}

func min2(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max2(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: src/features/typing/data/dart/median_of_two_sorted_arrays.dart
class Solution {
  double findMedianSortedArrays(List<int> nums1, List<int> nums2) {
    final m = nums1.length;
    final n = nums2.length;
    if (m > n) {
      return findMedianSortedArrays(nums2, nums1);
    }
    final total = m + n;
    final half = total ~/ 2;
    var left = 0;
    var right = m;
    while (left <= right) {
      final i = left + (right - left) ~/ 2;
      final j = half - i;
      final aLeft = i == 0 ? double.negativeInfinity : nums1[i - 1].toDouble();
      final aRight = i == m ? double.infinity : nums1[i].toDouble();
      final bLeft = j == 0 ? double.negativeInfinity : nums2[j - 1].toDouble();
      final bRight = j == n ? double.infinity : nums2[j].toDouble();
      if (aLeft <= bRight && bLeft <= aRight) {
        if (total % 2 == 1) {
          return aRight < bRight ? aRight : bRight;
        }
        final maxLeft = aLeft > bLeft ? aLeft : bLeft;
        final minRight = aRight < bRight ? aRight : bRight;
        return (maxLeft + minRight) / 2;
      }
      if (aLeft > bRight) {
        right = i - 1;
      } else {
        left = i + 1;
      }
    }
    return 0;
  }
}
